File: examples/testapp.py
#!/usr/bin/env python

import gobject
import dbus, dbus.glib

REPO_URL = "http://www.k-d-w.org/clipboard/deskbar/deskbar.xml"

CAPUCHIN_SERVICE = "org.gnome.Capuchin"
APPOBJECTMANAGER_PATH = "/org/gnome/Capuchin/AppObjectManager"
APPOBJECTMANAGER_IFACE = "org.gnome.Capuchin"
APPOBJECT_IFACE = "org.gnome.Capuchin.AppObject"

# We define some plugins here
PLUGINS = [('ekiga.py', '0.1.0'), # id, version
           ('yahoo.py', '0.3.0'),
           ('gmail-deskbar-hack.py', '0.99.9.99'),
          ]

class TestApp:
    
    def __init__(self):
        bus = dbus.SessionBus()
        # Get proxy object
        proxy_obj_manager = bus.get_object(CAPUCHIN_SERVICE, APPOBJECTMANAGER_PATH)
        # Apply the correct interace to the proxy object
        self.appobjectmanager = dbus.Interface(proxy_obj_manager, APPOBJECTMANAGER_IFACE)
        
        object_path = self.appobjectmanager.GetAppObject(REPO_URL)
        
        proxy_obj = bus.get_object(CAPUCHIN_SERVICE, object_path)
        self.appobject = dbus.Interface(proxy_obj, APPOBJECT_IFACE)
        
        self.appobject.connect_to_signal('InstallFinished', self.on_install_finished)
        self.appobject.connect_to_signal('Status', self.on_status)
        
    def on_install_finished(self, plugin_id):
        print "appobject updated: %s" % plugin_id
        
    def on_update_finished(self):
        self.testGetApplicationName();
        self.testGetAvailablePlugins();
        self.testGetAvailableUpdates();
        self.testGetTags ();
        self.testGetPluginsWithTag ();
        self.testInstall ();
        
    def on_install_finished(self, plugin_id):
        print "appobject updated: %s" % plugin_id
        self.appobject.Close() # We did everything we wanted to
	
    def on_status(self, action, plugin_id, progress, speed):
        print "DOWNLOAD: %s %s %s %s" % (action, plugin_id, progress, speed)
        
    def on_error(self, error):
        print "ERROR: " +error
        
    def testUpdate(self):
        self.appobject.Update(False, reply_handler=self.on_update_finished, error_handler=self.on_error)
        
    def testGetAvailablePlugins(self):
        stuff = self.appobject.GetAvailablePlugins();
        print "ALL:"
        for s in stuff:
            self.printPluginInfos (s)
			
    def printPluginInfos(self, s):
        print "ID: " + s
        print "Name: " + self.appobject.GetPluginName(s)
        print "Description: " + self.appobject.GetPluginDescription(s)
        author = self.appobject.GetPluginAuthor(s)
        print "Author: %s <%s>" % (author[0], author[1])
        tags = self.appobject.GetPluginTags(s);
        print "TAGS for %s:" % s
        for t in tags:
	        print t
        print
        
    def testGetApplicationName(self):
        print "APPLICATION-NAME: "+ self.appobject.GetApplicationName ()
        
    def testGetAvailableUpdates(self):
        updates = self.appobject.GetAvailableUpdates (PLUGINS)
        print "UPDATES:"
        for s in updates:
            print s
            print "Changes: "+self.appobject.GetPluginChanges(s, "1.1.0.0")
            
    def testGetTags(self):
        tags = self.appobject.GetTags ()
        print "TAGS:"
        for t in tags:
            print t
            
    def testGetPluginsWithTag(self):
        print "PLUGINS WITH TAG:"
        ids = self.appobject.GetPluginsWithTag ("phone")
        for i in ids:
            print self.printPluginInfos (i)
            
    def testInstall(self):
        self.appobject.Install("ekiga.py")
    
if __name__=='__main__':
    gobject.threads_init()
    dbus.glib.init_threads()
    mainloop = gobject.MainLoop()
    app = TestApp()
    app.testUpdate ()
    
    mainloop.run()

File: examples/testapp.cs
using System;
using System.Collections.Generic;
using System.Threading;
using NDesk.DBus;

public class TestCapuchin
{
	protected const string CAPUCHIN_SERVICE = "org.gnome.Capuchin";
	protected const string APPOBJECTMANAGER_PATH = "/org/gnome/Capuchin/AppObjectManager";
	
	protected Capuchin.IAppObjectManager manager;
	protected Capuchin.IAppObject appobject;
	
	private static TestCapuchin test = new TestCapuchin();
	
	public TestCapuchin()
	{
		Bus bus = Bus.Session;
		
		this.manager = bus.GetObject<Capuchin.IAppObjectManager> (CAPUCHIN_SERVICE, new ObjectPath (APPOBJECTMANAGER_PATH)); 
	}
	
	public void testGetAppObject()
	{
		ObjectPath path = this.manager.GetAppObject ("http://www.k-d-w.org/clipboard/deskbar/deskbar.xml");
		
		Bus bus = Bus.Session;
		
		this.appobject = bus.GetObject<Capuchin.IAppObject> (CAPUCHIN_SERVICE, path);
		this.appobject.InstallFinished += new Capuchin.InstallFinishedHandler ( this.OnInstallFinished );
		this.appobject.Status += new Capuchin.StatusHandler ( this.OnStatus );
	}
	
	protected void OnInstallFinished(string plugin_id)
	{
		Console.WriteLine ("appobject updated: {0}", plugin_id);
	}
	
	protected void OnStatus(Capuchin.ActionType action, string plugin_id, double progress, int speed)
	{
		Console.WriteLine ("DOWNLOAD: {0} {1} {2} {3}", action, plugin_id, progress, speed);
	}
        
	public void testUpdate()
	{
		this.appobject.Update(false);
	}
	
	public void testGetAvailablePlugins()
	{
		string[] stuff = this.appobject.GetAvailablePlugins();
		Console.WriteLine ("ALL:");
		foreach (string s in stuff)
		{
			this.printPluginInfos (s);
		}
	}
	
	private void printPluginInfos (string s) {
		Console.WriteLine ("ID: " + s);
        Console.WriteLine ("Name: " + this.appobject.GetPluginName(s));
        Console.WriteLine ("Description: " + this.appobject.GetPluginDescription(s) );
        string[] author = this.appobject.GetPluginAuthor(s);
        Console.WriteLine ("Author: {0} <{1}>", author[0], author[1]);
		string[] tags = this.appobject.GetPluginTags(s);
		Console.WriteLine ("TAGS for {0}:", s);
		foreach (string t in tags)
		{
			Console.WriteLine (t);
		}
		Console.WriteLine();
	}
	
	public void testGetAvailableUpdates()
	{
		string[][] plugins = new string[2][];
		plugins[0] = new string[] {"ssh.py", "0.0.9"};
		plugins[1] = new string[] {"ekiga.py", "0.1.0"};
		
		string[] updates = this.appobject.GetAvailableUpdates (plugins);
		Console.WriteLine ("UPDATES:");
		foreach (string s in updates)
		{
			Console.WriteLine (s);
            Console.WriteLine ("Changes: " + this.appobject.GetPluginChanges(s, "1.1.0.0"));
		}
	}
	
    public void testGetApplicationName () {
        Console.WriteLine ("APPLICATION-NAME: {0}", this.appobject.GetApplicationName ());
    }
	
	public void testGetTags () {
		string[] tags = this.appobject.GetTags ();
		Console.WriteLine ("TAGS:");
		foreach (string t in tags)
		{
			Console.WriteLine (t);
		}
	}
	
	public void testGetPluginsWithTag () {
		Console.WriteLine("PLUGINS WITH TAG:");
		string[] ids = this.appobject.GetPluginsWithTag ("phone");
		foreach (string id in ids) {
			this.printPluginInfos (id);
		}
	}
	
	public void testInstall()
	{
		this.appobject.Install("ekiga.py");
	}
	
	public void testClose()
	{
		this.appobject.Close();
	}
	
	public static void Main (string[] args)
	{
		test.testGetAppObject();
        test.testUpdate();
		test.testGetApplicationName();
        test.testGetAvailablePlugins();
		test.testGetAvailableUpdates();
		test.testGetTags ();
		test.testGetPluginsWithTag ();
		test.testInstall();
		Thread.Sleep(10000); // Wait 5s for update to complete, because we have no mainloop
		test.testClose();
	}
}
